File: Python/list/trapping_rain_water.py
import unittest
from typing import List
from unittest import TestCase

from parameterized import parameterized


def params():
    return [
        ([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1], 6),
        ([4, 2, 0, 3, 2, 5], 9),
        ([4, 2, 4], 2),
        ([0, 0, 0], 0),
        ([10, 20], 0),
        ([3, 3, 3, 3], 0),
        ([1, 2, 3, 4], 0),
        ([6, 5, 4, 3], 0),
        ([], 0),
        ([10], 0),
        (None, 0)
    ]


class Test(TestCase):

    @parameterized.expand(params())
    def test_trapping_water(self, given, expected):
        actual = trap(given)
        self.assertEqual(expected, actual)


def trap(height: List[int]) -> int:
    if height is None or len(height) < 3:
        return 0
    size = len(height)
    res = 0
    left_max = [None] * size
    right_max = [None] * size
    left_max[0] = height[0]
    right_max[size - 1] = height[size - 1]
    for i in range(1, size):
        left_max[i] = max(height[i], left_max[i - 1])
    for i in range(size - 2, -1, -1):
        right_max[i] = max(height[i], right_max[i + 1])
    for i in range(size):
        res += min(left_max[i], right_max[i]) - height[i]
    return res


if __name__ == '__main__':
    unittest.main()

File: Java/src/main/java/com/home/algorithms/leetcode/array/TrappingRainWater.java
package com.home.algorithms.leetcode.array;

public class TrappingRainWater {

    public int trap(int[] height) {
        if (height == null || height.length < 3) {
            return 0;
        }
        int size = height.length;
        int[] leftMax = new int[size];
        leftMax[0] = height[0];
        int[] rightMax = new int[size];
        rightMax[size - 1] = height[size - 1];
        for (int i = 1; i < size; i++) {
            leftMax[i] = Math.max(height[i], leftMax[i - 1]);
        }
        for (int i = size - 2; i >= 0; i--) {
            rightMax[i] = Math.max(height[i], rightMax[i + 1]);
        }
        int res = 0;
        for (int i = 0; i < size; i++) {
            res += Math.min(leftMax[i], rightMax[i]) - height[i];
        }
        return res;
    }
}
